src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -1,167 +1,373 @@
+      throw new CopycatException(e);
+    } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
+      throw new CopycatException("Reflection exception: ", e);
     }
   }
 
-
-  public Map<TopicPartition, Long> getPreviousOffsets() {
+  /**
+   * @return get committed offsets of previous task runs.
+   */
+  public Map<TopicPartition, Long> getCommittedOffsets() {
     return offsets;
   }
 
-  public void writeRecord(TopicPartition topicPart, SinkRecord record)
-      throws CopycatException{
-    try {
-      RecordWriter<Long, SinkRecord> writer = getWriter(topicPart, record);
-      writer.write(System.currentTimeMillis(), record);
-      if (!offsets.containsKey(topicPart)) {
-        offsets.put(topicPart, record.getOffset() - 1);
+  public void write(Collection<SinkRecord> records) throws IOException {
+    String topic;
+    int partition;
+    for (SinkRecord record: records) {
+      topic = record.topic();
+      partition = record.kafkaPartition();
+      TopicPartition topicPart = new TopicPartition(topic, partition);
+      buffer.get(topicPart).add(record);
+    }
+    for (TopicPartition topicPart: assignment) {
+      if (failureTime.containsKey(topicPart)) {
+        long now = System.currentTimeMillis();
+        if (now - failureTime.get(topicPart) < backOffMs) {
+          continue;
+        }
+        failureTime.remove(topicPart);
       }
-      updateRecordCounter(topicPart);
-      if (shouldRotate(topicPart)) {
-        rotate(topicPart);
+      execute(topicPart);
+    }
+  }
+
+  public void recover(TopicPartition topicPart) {
+    try {
+      switch (states.get(topicPart)) {
+        case RECOVERY_STARTED:
+          pause(topicPart);
+          nextState(topicPart);
+        case RECOVERY_PARTITION_PAUSED:
+          createWAL(topicPart);
+          nextState(topicPart);
+        case WAL_CREATED:
+          applyWAL(topicPart);
+          nextState(topicPart);
+        case WAL_APPLIED:
+          truncateWAL(topicPart);
+          nextState(topicPart);
+        case WAL_TRUNCATED:
+          resetOffsets(topicPart);
+          nextState(topicPart);
+        case OFFSET_RESET:
+          resume(topicPart);
+          nextState(topicPart);
+          break;
+        default:
+          log.error("{} is not a valid state to perform recovery.", states.get(topicPart));
       }
-    } catch (IOException e) {
-      throw new CopycatRuntimeException(e);
+    } catch (CopycatException e) {
+      log.error("Recovery failed.");
+      recordFailureTime(topicPart);
+      setRetryBackoff(backOffMs);
     }
   }
 
-  public void close() throws IOException, CopycatException {
-    for (TopicPartition topicPart: writers.keySet()) {
-      rotate(topicPart);
+  private void execute(TopicPartition topicPart) throws IOException {
+    Queue<SinkRecord> records = buffer.get(topicPart);
+    while(!records.isEmpty()) {
```

Comment:
I'm not sure recovery is guaranteed to finish with the way the code is now. `recover()` is only called in `start()`. If it fails part way through, the topic partition will be paused. This means that now records will be added to that topic partition's buffer. Therefore this loop will never be entered and the topic partition will effectively be stalled.